### PR TITLE
Add support for Checksum extern and CRC16/32 algorithms to eBPF backend

### PR DESCRIPTION
**Pull request by kmateuszssak**
This PR adds Checksum extern and adds support for CRC16 and CRC32 algorithms.

Co-authored-by: Tomasz Osiński <tomasz@opennetworking.org>
Co-authored-by: Jan Palimąka <jan.palimaka@orange.com>

### CODE REVIEW

File: backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        if (i & poly)
+            result |= j;
+    }
+
+    return Util::printf_format("%llu", result);
+}
+
+void CRCChecksumAlgorithm::emitUpdateMethod(CodeBuilder* builder, int crcWidth) {
+    // Note that this update method is optimized for our CRC16 and CRC32, custom
+    // version may require other method of update. To deal with byte order data
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp:67]: comma after order?

File: backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            ::error(ErrorType::ERR_UNSUPPORTED, "Only bits types are supported %1%", field);
+            return;
+        }
+        const int width = fieldType->width_bits();
+
+        if (width < 8 || concatenateBits) {
+            concatenateBits = true;
+            if (width > remainingBits) {
+                ::error(ErrorType::ERR_UNSUPPORTED,
+                        "Sub-byte fields have to be aligned to bytes %1%", field);
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp:154]: I don't think I would understand this error message as a user of the compiler.

File: backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    builder->endOfStatement(true);
+}
+
+void CRCChecksumAlgorithm::emitClear(CodeBuilder* builder) {
+    builder->emitIndent();
+    builder->appendFormat("%s = %s", registerVar.c_str(), initialValue.c_str());
+    builder->endOfStatement(true);
+}
+
+void CRCChecksumAlgorithm::emitAddData(CodeBuilder* builder,
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp:129]: This code has several layers of indirection for generating a string that becomes C code, so it's not easy to read.
Can you add a comment showing some final generated code?

File: backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.h [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    void emitGet(CodeBuilder* builder) override;
+
+    void emitSubtractData(CodeBuilder* builder, const argumentsList & arguments) override;
+
+    void emitGetInternalState(CodeBuilder* builder) override;
+    void emitSetInternalState(CodeBuilder* builder,
+                              const IR::MethodCallExpression * expr) override;
+};
+
+class CRC16ChecksumAlgorithm : public CRCChecksumAlgorithm {
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.h:107]: Perhaps you can add a comment describing how this protocol is used to compute the checksum: how these values are used, and what is the order that these methods are called in the generated code.

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]
This is good, but could use a little more documentation.

File: backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+void EBPFHashAlgorithmPSA::emitAddData(CodeBuilder *builder, int dataPos,
+                                       const IR::MethodCallExpression *expr) {
+    emitAddData(builder, unpackArguments(expr, dataPos));
+}
+
+// ===========================CRCChecksumAlgorithm===========================
+
+cstring CRCChecksumAlgorithm::reflect(cstring str) {
```
[the remark this answers is not in the document]
Comment (kmateuszssak) [reply backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp]: I removed this method and commented why `polynomial` has a such value.

File: backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+EBPFHashAlgorithmPSA::argumentsList EBPFHashAlgorithmPSA::unpackArguments(
+        const IR::MethodCallExpression * expr, int dataPos) {
+    BUG_CHECK(expr->arguments->size() > ((size_t) dataPos),
+              "Data position %1% is outside of the arguments: %2%", dataPos, expr);
+
+    std::vector<const IR::Expression *> arguments;
+
+    if (expr->arguments->at(dataPos)->expression->is<IR::StructExpression>()) {
+        auto argList = expr->arguments->at(dataPos)->expression->to<IR::StructExpression>();
```
[the remark this answers is not in the document]
Comment (kmateuszssak) [reply backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp:31]: If expression is not a StructExpression then then this expression is added to the arguments vector. So there is no any error.

File: backends/ebpf/tests/p4testdata/checksum-updates-crc16.p4 [hunk 7]
```diff
@@ -0,0 +1,133 @@
+/*
```
[the remark this answers is not in the document]
Comment (kmateuszssak) [reply backends/ebpf/tests/p4testdata/checksum-updates-crc16.p4:1]: I think we shouldn't. What would be the value of adding PTF testfiles to `testdata/p4_16_samples` ?

**Comment by kmateuszssak** [comment]
@mbudiu-vmw I think I've addressed all of the comments.

File: backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        for (auto field : argList->components)
+            arguments.push_back(field->expression);
```
Comment (osinstom) [inline backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp:33]: @kmateuszssak we could eventually simplify that as follows:

```suggestion
    if (auto argList = expr->arguments->at(dataPos)->expression->to<IR::StructExpression>()) {
        for (auto field : argList->components)
            arguments.push_back(field->expression);
```

**Review by mbudiu-vmw — APPROVED** [review]

**Comment by mbudiu-vmw** [comment]
Unfortunately this branch now has conflicts, which you will have to resolve before we can merge.

**Pull request merged**